### PR TITLE
Data management: dpc++ API improvements

### PR DESCRIPTION
**Pull request by michael-smirnov**
Changes:
- `data_parallel_policy` - implementation changed not to create pimpl for naive construction from `sycl::queue` (perf improvement)
- `host_policy` avoided in DM APIs for performance and priority reasons; internally, dispatching is organized by `default_host_policy`
- all policies moved to details
- `host_only_alloc` removed, allocators added to details
- `array` - explicit constructors, `ConstDeleter` typename in appropriate places
- `array` - `wrap()` methods implemented for const and non-const data
- `array` - API of constructor accepting USM memory changed - queue is deleted, deleter added
- `array_impl` created, `array` now holds pimpl and hot data pointers 
- `homogen_table` - USM constructor now accepts a deleter.
- `homogen_table` constructor - pointer placed first.

### CODE REVIEW

File: cpp/oneapi/dal/data/array_dpc_test.cpp [hunk 1]
```diff
@@ -108,17 +108,36 @@ TEST(array_dpc_test, can_reset_array_with_raw_pointer) {
...
     constexpr int64_t count = 10;
     auto* data              = sycl::malloc_shared<float>(count, q);
-    auto event              = q.submit([&](sycl::handler& cgh) {
-        cgh.parallel_for(sycl::range<1>(count), [=](sycl::id<1> idx) {
-            data[idx[0]] = idx;
-        });
-    });
+    q.submit([&](sycl::handler& cgh) {
+         cgh.parallel_for(sycl::range<1>(count), [=](sycl::id<1> idx) {
+             data[idx[0]] = idx;
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/data/array_dpc_test.cpp:113]: Do you mean 
data[idx[0]] = static_cast<float>(idx[0]);

Comment (michael-smirnov) [reply cpp/oneapi/dal/data/array_dpc_test.cpp]: yes, thanks

**Pull request merged**